src/world/Manager/MapMgr.cpp

```diff
@@ -0,0 +1,889 @@
+        {
+          auto pNpc = reinterpret_cast< LGB_ENPC_ENTRY* >( pEntry.get() );
+
+          auto eNpcData = exdData.get< Data::ENpcBase >( pNpc->data.enpcId )->eNpcData;
+
+          for( auto npcData : eNpcData )
+          {
+            if( npcData == 0 )
+              continue; // Some npcs have data gaps, so we have to iterate through the entire array
+
+            if( count++ % 1000 == 0 )
+              std::cout << ".";
+
+            EventData eventData;
+            eventData.levelId = pNpc->data.instanceId;
+            eventData.actorId = pNpc->data.enpcId;
+
+            auto eventHandlerType = static_cast< EventHandler::EventHandlerType >( npcData >> 16 );
+
+            switch( eventHandlerType )
+            {
+              case EventHandler::EventHandlerType::GuildLeveAssignment:
+              {
+                eventData.iconId = exdData.get< Data::EventIconType >( 5 )->mapIconAvailable + 1;
+
+                m_mapData[ territoryTypeId ].insert( std::make_pair( npcData, eventData ) );
+                break;
+              }
+
+              case EventHandler::EventHandlerType::CustomTalk:
+              {
+                // Include only the beginner arena icon yet. There a few other ones, that aren't referenced in the game files (Some examples are: The Triple Triad Tournament npc which has multiple icons and the ocean fishing icon)
+                if( npcData == 721223 )
+                {
+                  auto customTalk = exdData.get< Data::CustomTalk >( npcData );
+
+                  eventData.iconId = customTalk->iconMap;
+
+                  m_mapData[ territoryTypeId ].insert( std::make_pair( npcData, eventData ) );
+                }
+                break;
+              }
+
+              case EventHandler::EventHandlerType::GuildOrderGuide:
+              {
+                eventData.iconId = exdData.get< Data::EventIconType >( 6 )->mapIconAvailable + 1;
+
+                m_mapData[ territoryTypeId ].insert( std::make_pair( npcData, eventData ) );
+                break;
+              }
+
+              case EventHandler::EventHandlerType::TripleTriad:
+              {
+                eventData.iconId = exdData.get< Data::EventIconType >( 7 )->mapIconAvailable + 1;
+
+                m_mapData[ territoryTypeId ].insert( std::make_pair( npcData, eventData ) );
+                break;
+              }
+
+              case EventHandler::EventHandlerType::PreHandler:
+              {
+                //I think this is used in Bozja and Zadnor, need evidence
+                //m_mapData[ territoryTypeId ].insert( std::make_pair( eventData, 0 ) );
+                break;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  std::cout << "\n";
+}
+
+void Sapphire::World::Manager::MapMgr::updateAll( Entity::Player& player )
+{
+  auto& mapData = m_mapData[ player.getTerritoryTypeId() ];
+  std::multimap< uint32_t, EventData, less > newMapData;
```

Comment:
is there a particular reason for using a multimap instead of just a list with a defined structure?

Reply:
The retail server seems to sort it by the lower word of the event id.
So I used a multimap for auto sorting with a custom sort method.
For lower processing time we can use a list here, because it seems the sorting is irrelevant for the client.

Reply:
yeah my main concern was that this obscures the behavior and makes it a bit more complicated for no real gain - if ordering was necessary though i would just opt to store them with the event id as they'll be implicitly ordered by that anyway if they're dumped into a set/map when the key is the event id